### PR TITLE
Check for symlink creation privilege for portable install

### LINKED ISSUE #2364: Installer failed with exit code: 0x8a150003
### Brief description of your issue

I was attempting to install Microsoft.NuGet (a portable package). This package failed to install in user mode, but installed correctly in administrator mode.

*Note:* This package was added to the repository as a reference example for a portable package.

### Steps to reproduce

`winget install Microsoft.NuGet`

### Expected behavior

Install Succeeds

### Actual behavior

```PowerShell
winget install Microsoft.NuGet
Found NuGet [Microsoft.NuGet] Version 6.2.1.2
This application is licensed to you by its owner.
Microsoft is not responsible for, nor does it grant any licenses to, third-party packages.
Downloading https://dist.nuget.org/win-x86-commandline/v6.2.1/nuget.exe
  ██████████████████████████████  6.75 MB / 6.75 MB
Successfully verified installer hash
Starting package install...
Overwriting existing file:C:\Users\denelon\AppData\Local\Microsoft\WinGet\Links\nuget.exe
An unexpected error occurred while executing the command:
create_symlink: A required privilege is not held by the client.: "C:\Users\denelon\AppData\Local\Microsoft\WinGet\Packages\Microsoft.NuGet_Microsoft.Winget.Source_8wekyb3d8bbwe\nuget.exe", "C:\Users\denelon\AppData\Local\Microsoft\WinGet\Links\nuget.exe"
Portable install failed; Cleaning up...
Starting package uninstall...
Installer failed with exit code: 0x8a150003 : Executing command failed
```

### Environment

```shell
Windows Package Manager (Preview) v1.4.2011-preview
Copyright (c) Microsoft Corporation. All rights reserved.

Windows: Windows.Desktop v10.0.22000.832
System Architecture: X64
Package: Microsoft.DesktopAppInstaller v1.19.2011.0
```

### PR DESCRIPTION
**Pull request by ryfu-msft**
Fixes #2364 
Related to #2368 

Checks if developer mode is enabled for portable install. If developer mode is not enabled, requires command to be running as admin. Additionally, a link to the related GitHub issue is displayed to the user. 

Tests:
- Added test functionality to verify IsDevModeEnabled()
- Added unit test to verify that github issue is displayed properly if developer mode is off and not running as admin.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/2369)

### CODE REVIEW

File: src/AppInstallerCLICore/Workflows/PortableFlow.cpp [hunk 1]
```diff
@@ -540,8 +541,23 @@ namespace AppInstaller::CLI::Workflow
...
+                // TODO: Remove once issue regarding symlink creation privilege has been resolved.
+                if (context.IsTerminated())
+                {
+                    context.Reporter.Error() << std::endl << "https://github.com/microsoft/winget-cli/issues/2368" << std::endl;
+                }
+            }
+        }
     }
-
+
```
Comment (Trenly) [inline src/AppInstallerCLICore/Workflows/PortableFlow.cpp:560]: nit: whitespace

File: src/AppInstallerCommonCore/Runtime.cpp [hunk 2]
```diff
@@ -704,6 +707,21 @@ namespace AppInstaller::Runtime
...
     }
 
+    // Determines whether developer mode is enabled.
+    bool IsDevModeEnabled()
+    {
+        const auto& devModeSubKey = Registry::Key::OpenIfExists(HKEY_LOCAL_MACHINE, s_DevModeSubkey, 0, KEY_READ|KEY_WOW64_64KEY);
+        const auto& devModeEnabled = devModeSubKey[s_AllowDevelopmentWithoutDevLicense];
+        if (devModeEnabled.has_value())
+        {
+            return devModeEnabled->GetValue<Registry::Value::Type::DWord>();
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/Runtime.cpp:717]:
```suggestion
            return devModeEnabled->GetValue<Registry::Value::Type::DWord>() == 1;
```
The only value that enables it is 1.

File: src/AppInstallerCLITests/TestCommon.cpp [hunk 3]
```diff
@@ -220,6 +220,13 @@ namespace TestCommon
         THROW_IF_WIN32_ERROR(RegSetValueExW(key, name.c_str(), 0, REG_DWORD, reinterpret_cast<const BYTE*>(&value), sizeof(DWORD)));
     }
 
+    void EnableDevMode(bool enable)
+    {
+        wil::unique_hkey result;
+        THROW_IF_WIN32_ERROR(RegOpenKeyExW(HKEY_LOCAL_MACHINE, L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\AppModelUnlock", 0, KEY_ALL_ACCESS|KEY_WOW64_64KEY, &result));
+        SetRegistryValue(result.get(), L"AllowDevelopmentWithoutDevLicense", enable);
```
Comment (JohnMcPMS) [inline src/AppInstallerCLITests/TestCommon.cpp:227]:
```suggestion
        SetRegistryValue(result.get(), L"AllowDevelopmentWithoutDevLicense", (enable ? 1 : 0));
```
The only value that enables it is 1.

File: src/AppInstallerCommonCore/Runtime.cpp [hunk 4]
```diff
@@ -704,6 +707,21 @@ namespace AppInstaller::Runtime
         return wil::test_token_membership(nullptr, SECURITY_NT_AUTHORITY, SECURITY_BUILTIN_DOMAIN_RID, DOMAIN_ALIAS_RID_ADMINS);
     }
 
+    // Determines whether developer mode is enabled.
+    bool IsDevModeEnabled()
+    {
+        const auto& devModeSubKey = Registry::Key::OpenIfExists(HKEY_LOCAL_MACHINE, s_DevModeSubkey, 0, KEY_READ|KEY_WOW64_64KEY);
```
Comment (JohnMcPMS) [inline src/AppInstallerCommonCore/Runtime.cpp:713]: There is also a group policy value that controls it and takes precedence over the user one.  For this temporary use case this is good enough, but might be worth commenting on it.

**Review by JohnMcPMS — CHANGES_REQUESTED** [review]

File: src/AppInstallerCLICore/Workflows/PortableFlow.cpp (hunk 1 above)
Comment (ryfu-msft) [reply src/AppInstallerCLICore/Workflows/PortableFlow.cpp]: Removed whitespace

File: src/AppInstallerCommonCore/Runtime.cpp (hunk 2 above)
Comment (ryfu-msft) [reply src/AppInstallerCommonCore/Runtime.cpp]: Changed to suggested

File: src/AppInstallerCLICore/Workflows/PortableFlow.cpp [hunk 5]
```diff
@@ -540,8 +541,23 @@ namespace AppInstaller::CLI::Workflow
...
         }
+
+        void EnsureSymlinkCreationPrivilege(Execution::Context& context)
+        {
+            if (!Runtime::IsDevModeEnabled())
+            {
+                AICLI_LOG(CLI, Info, << "Developer mode not enabled.");
+                context << Workflow::EnsureRunningAsAdmin;
+
+                // TODO: Remove once issue regarding symlink creation privilege has been resolved.
```
[the remark this answers is not in the document]
Comment (ryfu-msft) [reply src/AppInstallerCLICore/Workflows/PortableFlow.cpp]: Clarified TODO comment

File: src/AppInstallerCLITests/TestCommon.cpp (hunk 3 above)
Comment (ryfu-msft) [reply src/AppInstallerCLITests/TestCommon.cpp]: changed to suggested

File: src/AppInstallerCommonCore/Runtime.cpp (hunk 4 above)
Comment (ryfu-msft) [reply src/AppInstallerCommonCore/Runtime.cpp:714]: Added a comment clarifying the group policy behavior effect on developer mode setting

**Review by JohnMcPMS — APPROVED** [review]

**Comment by jantari** [comment]
Why does this test for running as admin? AFAIK all that is required to create symlinks is having SeCreateSymbolicLinkPrivilege, not being an administrator.

**Comment by denelon** [comment]
@jantari it requires either the user to be in administrator mode or to have developer mode enabled on Windows. A normal user account otherwise can't create symbolic links. It was a very late breaking issue we ran into.

**Pull request merged**